### PR TITLE
BAAS-22177: add mem usage estimate for sparse arrays

### PR DESCRIPTION
**Pull request by Gabri3l**
Adding the SizeUint32 was probably unnecessary =D

### CODE REVIEW

File: array_sparse.go [hunk 1]
```diff
@@ -512,9 +539,23 @@ func (a *sparseArrayObject) MemUsage(ctx *MemUsageContext) (memUsage uint64, err
 	// sparseArrayObject overhead
 	memUsage = SizeEmptyStruct
 
+	inc, err := a.baseObject.MemUsage(ctx)
```
Comment (kpatel71716) [inline array_sparse.go:542]: [q] do we need to check the err here?

File: array_sparse.go [hunk 2]
```diff
@@ -499,6 +499,33 @@ func (a *sparseArrayObject) exportToArrayOrSlice(dst reflect.Value, typ reflect.
...
+		// We add both the mem usage of the value and its index
+		memUsage += inc + SizeUint32
+		if err != nil {
+			return computeMemUsageEstimate(memUsage, samplesVisited, totalItems), err
+		}
+	}
+
+	return computeMemUsageEstimate(memUsage, samplesVisited, totalItems), nil
+}
+
```
Comment (kpatel71716) [inline array_sparse.go:528]: [q] alluding to Nick's point on [this](https://github.com/mongodb-forks/goja/pull/120/files) PR, are we intentionally using different logic for sparse arrays and maps?

**Review by kpatel71716 — COMMENTED** [review]
Looks pretty good, couple of small qs before approving

File: array_sparse.go (hunk 1 above)
Comment (Gabri3l) [reply array_sparse.go:542]: oh wow yes we do!

File: array_sparse.go [hunk 3]
```diff
@@ -499,6 +499,33 @@ func (a *sparseArrayObject) exportToArrayOrSlice(dst reflect.Value, typ reflect.
...
+		return memUsage, nil
+	}
+	sampleSize := totalItems / 10
+
+	// grabbing one sample every "sampleSize" to provide consistent
+	// memory usage across function executions
+	for i := 0; i < totalItems; i += sampleSize {
+		v := a.items[i].value
+		inc, err := v.MemUsage(ctx)
+		samplesVisited += 1
```
Comment (nickpoindexter) [inline array_sparse.go:518]: [q] Can `samplesVisited` just be total items? The only time it's not is on error which I imagine we care less about.

**Review by nickpoindexter — APPROVED** [review]
LGTM. Just one minor improvement.

File: array_sparse.go [hunk 4]
```diff
@@ -512,9 +539,23 @@ func (a *sparseArrayObject) MemUsage(ctx *MemUsageContext) (memUsage uint64, err
...
+	memUsage += inc
+
+	if ctx.ArrayLenExceedsThreshold(len(a.items)) {
+		inc, err := a.estimateMemUsage(ctx)
+		memUsage += inc
+		if err != nil {
+			return memUsage, err
+		}
+
+		ctx.Ascend()
```
Comment (jwongmongodb) [inline array_sparse.go:555]: [q] is the ctx.Ascend and Descend used as a stack to calculate the mem usage for every items for each level? How come we need to explicitly put ctx.Ascend() here? Is it better to do a `defer ctx.Ascend` after the descend so it pops out if there are any errors?

Comment (Gabri3l) [reply array_sparse.go:555]: You make a good point but I'd rather make this change separately so it's easy to revert.

File: array_sparse.go (hunk 2 above)
Comment (Gabri3l) [reply array_sparse.go:528]: yeah the way data is stored is different in the two types so I have to adjust how we calculate memory accordingly. You can see the difference is reflected in the respective mem usage functions.

File: array_sparse.go (hunk 3 above)
Comment (Gabri3l) [reply array_sparse.go:518]: It cannot as given for example 100 total items, we sample every 10 so the samples visited would be 10 in this case.

**Review by jwongmongodb — APPROVED** [review]
LGTM, calling out that there may be a separate commit for the defer ascend case.

File: array_sparse.go (hunk 4 above)
Comment (Gabri3l) [reply array_sparse.go:555]: Created https://jira.mongodb.org/browse/BAAS-32009

File: array_sparse.go (hunk 3 above)
Comment (nickpoindexter) [reply array_sparse.go:518]: oh I see you're not incrementing by 1. Got it.

File: array_sparse.go (hunk 2 above)
Comment (kpatel71716) [reply array_sparse.go:528]: makes sense!

**Review by kpatel71716 — APPROVED** [review]
LGTM!

**Pull request merged**